src/plugins/share/public/persistable_state/types.ts

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { SavedObjectAttributes, SavedObjectReference } from 'src/core/public';
+
+export interface PersistableState<
+  S extends SavedObjectAttributes,
+  I extends string | undefined = undefined,
+  MS extends SavedObjectAttributes | undefined = undefined
+> {
+  state: S;
+  migratedId?: I;
+  migratedState?: MS;
+}
+
+/**
+ * Anybody registering persistable state should add a property to this
+ * interface so that any module importing PersistableStates can
+ * have the interfaces of all persistable states available to them.
+ *
+ *  declare module '../../plugins/share/public' {
+ *    interface PersistableStates {
+ *      myId: PersistableState<myState, myMigratedId, myMigratedState>;
+ *    }
+ *  }
+ *
+ * @public
+ */
+export interface PersistableStates {
+  [key: string]: PersistableState<SavedObjectAttributes>;
+}
+
+interface SavedObjectWithReferences {
+  attributes: SavedObjectAttributes;
+  references: SavedObjectReference[];
+}
+type ExtractReferences = (opts: SavedObjectWithReferences) => SavedObjectWithReferences;
+type InjectReferences = (opts: SavedObjectWithReferences) => void;
```

Comment:
should accept saved object + array of reference and return new object with injected references